### ISSUE TITLE
Gradient accumulation and batch size handling in training loop

### ISSUE THREAD

**Issue by yashsavani** [yifan123/flow_grpo#73]
I'm a bit confused about the gradient accumulation and batch size handling in the training loop, and I think there might be an issue with the rebatching logic.

### Main Issue:

At line 840-843 of `scripts/train_sd3.py`, the code rebatches samples using:
```python
samples_batched = {
    k: v.reshape(-1, total_batch_size//config.sample.num_batches_per_epoch, *v.shape[1:])
    for k, v in samples.items()
}
```

However, `config.sample.num_batches_per_epoch` is a sampling parameter, not a training parameter. For training, shouldn't this divide by something related to `config.train.gradient_accumulation_steps` instead? 

The Accelerator is configured with `gradient_accumulation_steps * num_train_timesteps`, so I believe the rebatching should align with the gradient accumulation configuration.

### Additional Confusion:

What's the intended difference between `config.train.batch_size` and `config.sample.train_batch_size`? Currently, `config.train.batch_size` is only used for creating negative prompt embeddings but doesn't seem to affect the actual training batch size.

### Possible Fix:

Perhaps line 840-843 should use:
```python
samples_batched = {
    k: v.reshape(-1, total_batch_size//config.train.gradient_accumulation_steps, *v.shape[1:])
    for k, v in samples.items()
}
```

Would appreciate clarification on the intended behavior here. Thanks!

**Comment by yifan123** [comment]
Could you provide the config where the above precision error occurred? I’ll take a look.

**Comment by yashsavani** [comment]
Thanks for the clarification earlier, but I’m still trying to pinpoint exactly why the precision differences from changing batch size between sampling and training cause reward maximization to fail.

On A100 80GB (bf16/fp16) I’m limited to a per-GPU batch size of 4 for training. This works for reward learning if I also sample with batch size 4 — but it leaves a lot of unused capacity during sampling. If I instead sample with a larger batch (to use the GPU more efficiently) and then use gradient accumulation in training to match the effective batch size, the reward drops off sharply.

I’d like to understand what’s happening numerically that makes this mismatch harmful for RL? Is it purely that re-running the model on the same states with a different per-forward microbatch changes the fp16/bf16 kernel choice/reduction order, causing small log-prob differences that break the ratio calculation? Or is there some other source of batch-size sensitivity here?

If the main issue is the ratio computation becoming noisy when `log_prob_new` is computed with a different microbatch size than `log_prob_old`, are there any possible fixes, such as:
- Caching log_prob_old from sampling and using it directly in training (avoiding re-running the old policy entirely)
- Forcing identical microbatch shapes and deterministic kernels during both sampling and the ratio pass in training
- Adjusting the RL loss to be more tolerant to very small ratio drift

Gradient accumulation would let me fully utilize the hardware during sampling while keeping training stable, so any approach to make this work would be a big win.
